feat: added reputation method to release

CzarekDryl opened the pull request:
## Description

- Added reputation method to release step in advanced payment

## Testing

* Install Reputation Weighted extension
* Create advanced payment and go to release step
* Choose reputation method from release modal

Main issue: [https://github.com/JoinColony/colonyCDapp/issues/2256](https://github.com/JoinColony/colonyCDapp/issues/2256)

iamsamgibbs commented:
@CzarekDryl Can you rebase this so I can give it a review?

CzarekDryl commented:
@iamsamgibbs Branch rebased

jakubcolony reviewed (changes requested):
Thanks for your PR @CzarekDryl, nice to see this coming together.

The message ID for Reputation in the release modal seems incorrect and causing errors:

<img width="504" alt="image" src="https://github.com/user-attachments/assets/448e8df2-08d4-4bb2-b0e7-533c7689a2eb">

<img width="554" alt="image" src="https://github.com/user-attachments/assets/96fa5280-5ffa-4f9f-9779-9922253f2f4e">

----

After releasing via a motion, the UI says it was the Payment creator decision method:
<img width="381" alt="image" src="https://github.com/user-attachments/assets/aca675ca-fc61-4991-a3b4-35bc8fea775b">

The rest of the flow looks good, including voting and failed motions, nice work 👍 

<img width="393" alt="image" src="https://github.com/user-attachments/assets/6d320738-55ff-4b9e-b585-f735546e35f2">

arrenv reviewed (changes requested):
@CzarekDryl Thank you for adding this in. I picked up

1. The text is wrong in the info callout, it should be "for releasing the payment."
![image](https://github.com/user-attachments/assets/66c60637-9a68-4538-ab1e-de095cc0b392)

2. Is there a way to not have the "Failed" show here before the "Passed"? Can we use the skeleton loader?

[failed-shows.webm](https://github.com/user-attachments/assets/5ffdefb6-dd2d-4edc-93b7-bc352e4d85f9)

CzarekDryl commented:
@arrenv 

1. Description changed to `for releasing the payment.`
2. We have the values from the beginning, only over time the value of yayVotes changes. The loading of the action is set to false in this case.
`{
    "nayStakes": "4426661121689491626",
    "requiredStake": "4426661121689491626",
    "yayStakes": "4426661121689491626",
    "yayVotes": "0",
    "nayVotes": "0"
} before

{
    "nayStakes": "4426661121689491626",
    "requiredStake": "4426661121689491626",
    "yayStakes": "4426661121689491626",
    "yayVotes": "398798442899037963861",
    "nayVotes": "0"
} after`

arrenv commented:
> 2. We have the values from the beginning, only over time the value of yayVotes changes. The loading of the action is set to false in this case.
> 
> ```
> {
>     "nayStakes": "4426661121689491626",
>     "requiredStake": "4426661121689491626",
>     "yayStakes": "4426661121689491626",
>     "yayVotes": "0",
>     "nayVotes": "0"
> } 
> ```
> 
> before
> 
> ```
> {
>     "nayStakes": "4426661121689491626",
>     "requiredStake": "4426661121689491626",
>     "yayStakes": "4426661121689491626",
>     "yayVotes": "398798442899037963861",
>     "nayVotes": "0"
> }
> ```
> 
> after

@CzarekDryl does this need to be an arbitrary setTimeout then, maybe like 3 seconds? That way, it allows time for the loading without showing a false outcome, but will still fallback to the hopefully rare occasions where a motion that gets to a vote does not receive any votes.

jakubcolony commented:
@arrenv the explanation from @CzarekDryl made me think this must be a race condition that also affects regular motions (as it's using the same code). After 3 attempts I managed to reproduce it on a mint tokens motion:

https://github.com/user-attachments/assets/1b142b03-4e20-466b-8d47-a8a849df5adf

Additionally the voting step shows the wrong outcome:

<img width="396" alt="image" src="https://github.com/user-attachments/assets/5fe2c96d-0803-48e9-95be-a3a6042a2fe3">

I have opened an issue #3390.

CzarekDryl commented on src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/hooks.tsx:
```diff
@@ -0,0 +1,192 @@
+    ) {
+      setExpectedStepKey(ExpenditureStep.Release);
+    }
+  }, [allReleaseMotions, isAnyReleaseMotionInProgress, setExpectedStepKey]);
```
It seems to me that I handled some case, but without it, everything also seems to work well

CzarekDryl commented on src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/ReleaseRequests/ReleaseRequests.tsx:
```diff
@@ -0,0 +1,37 @@
+import React, { type FC } from 'react';
```
I've added new shared component and changed `FundingRequests` and `FinalizeRequests` but left it in `ReleaseActions` because there is some additional logic there

jakubcolony reviewed (changes requested):
Thanks for your work @CzarekDryl, it is looking much better 👍 

Just a few things I ran into:

When you create another payment after finalizing the previous, the action/motion info still shows in the new payment (even though there aren't any finalize actions yet):

<img width="888" alt="image" src="https://github.com/user-attachments/assets/dfbe9734-2f9f-4f53-95e4-5a9c1c6f69be">

I remember having to deal with this when rebasing funding using motions, there's a cleanup useEffect in `PaymentBuilderWidget` that sets the selected funding action to null.

---- 

After finalizing a passed motion, the pills next to each payment say "Claim in" and never update. They should say "Claim now" if the claim delay is 0:

<img width="1078" alt="image" src="https://github.com/user-attachments/assets/51693be5-7f18-48f5-b993-189e0c134459">

----

Not sure if it's this PR that impacted it, but the Pending button is broken in the payment step:
<img width="390" alt="image" src="https://github.com/user-attachments/assets/9f96a98e-5161-4c24-8be4-c94c52e65c91">

----

The rest works fine:

<img width="407" alt="image" src="https://github.com/user-attachments/assets/5ff5b3f7-6728-4cab-9171-19b49d8c3515">
<img width="390" alt="image" src="https://github.com/user-attachments/assets/a31643e1-3da3-4b05-a6dc-c6a750ea2dfa">


<img width="396" alt="image" src="https://github.com/user-attachments/assets/833c3477-2f6c-4bd8-9dea-1650a51501a1">

<img width="397" alt="image" src="https://github.com/user-attachments/assets/6fce499f-2203-4278-8539-07af217e9a47">

CzarekDryl commented:
@jakubcolony I've finalized a passed motion and pill have changes to `claim now`
![image](https://github.com/user-attachments/assets/c51898d8-0fb4-42a9-b064-bd9bec7bdc54)

jakubcolony reviewed (commented):
@CzarekDryl did you also forward the time (and possibly refreshed the page) after finalizing? Considering the payout has a claim delay of 1 hour

I'm still able to reproduce it on every finalization: 

https://github.com/user-attachments/assets/13cc175c-c7f2-4fa5-b6df-231d305067e9

jakubcolony reviewed (changes requested):
I think it's slightly different with permissions in a way that it eventually changes to "Claimed", whereas here it will stay on "Claim in" until page refresh. Perhaps @arrenv could weigh in whether it's fine to merge like that and handle the issue separately.

Otherwise I'm happy to approve, except the small problem I encountered after the rebase. The extension has definitely not been uninstalled:

<img width="393" alt="image" src="https://github.com/user-attachments/assets/25e10349-7f1f-4d20-b72d-40a1aaa4c402">

arrenv commented:
@CzarekDryl & @jakubcolony It it certainly good to avoid having to reload for the UI to update, so, to avoid missing this, I would prefer to resolve it in this PR.

CzarekDryl commented:
@jakubcolony I've changed uninstalled extension condition as you said and also fixed claim in badge (please check if this approach is acceptable)

https://github.com/user-attachments/assets/366c78d3-87fb-49d4-ba56-5f31fa432c3b

jakubcolony reviewed (approved):
Yep, that has done the trick:

https://github.com/user-attachments/assets/76f47ef0-b6d0-47a9-b205-093fe0a82fb6

Nice work @CzarekDryl, this looks good to be merged now 👍

iamsamgibbs reviewed (changes requested):
Nice work on this so far! Most of the changes are looking great, and I am able to get through the entire flow.

I did run into a few issues whilst testing though.

I keep getting this console error popping up:

<img width="424" alt="Screenshot 2024-11-13 at 13 27 52" src="https://github.com/user-attachments/assets/8dfc1b0d-25df-40e7-9793-1bb02bc918a7">

It seems to be caused by the `FinalizePaymentModal` `onSuccess` function which runs when after submitting the form in the modal. I added a console log here to demonstrate (after 30 seconds the console error would pop up).

https://github.com/user-attachments/assets/0dfd4d99-aa8d-48b1-9da0-389a95bc5ae2

I'm guessing the intention here is for the `setIsWaitingForClaimedPayouts` function to run when we are on the Payment step to fetch the correct payout state. But because we are now using the reputation decision method, it doesn't instantly switch to the Payment step. So the onSuccess handler should only trigger for the permissions / payment creator decision methods. (This might also mean `setIsWaitingForClaimedPayouts` should be called elsewhere when the motion is finalised?)

I ran into this issue a few times when failing a motion, after finalizing the rejection, the "Release" button was stuck in a pending state.

<img width="1045" alt="Screenshot 2024-11-13 at 13 35 29" src="https://github.com/user-attachments/assets/357822dc-2492-450d-8b65-07cf1f554c11">

https://github.com/user-attachments/assets/dada57f4-a614-4a1d-a898-92643e9fc10e

The reveal stage failed to update a few times too and I had to refresh to be able to reveal my vote:

<img width="1037" alt="Screenshot 2024-11-13 at 13 41 25" src="https://github.com/user-attachments/assets/78f86a6b-3eb4-40fe-8323-1f1864957dbb">

As well as my rejection vote not triggering it to proceed to the next stage:

<img width="1303" alt="Screenshot 2024-11-13 at 13 44 01" src="https://github.com/user-attachments/assets/f77a1f8e-8040-46d1-a819-8e59d930adcf">

This will probably need someone on our side to look into as a separate issue, but the notification messages for all the motions related to the expenditure release are completely blank:

<img width="763" alt="Screenshot 2024-11-13 at 13 28 59" src="https://github.com/user-attachments/assets/4b1941cf-14e4-45e2-a5cb-838a2eafa4bd">

CzarekDryl commented:
@iamsamgibbs The reveal step should no longer freeze, and I set `setIsWaitingForClaimedPayouts` when the motion passes.
@arrenv I can get through all steps without reloading page, same as Sam.

iamsamgibbs reviewed (changes requested):
Hmm I'm still experiencing most of the same issues, including the console error.

<img width="1408" alt="Screenshot 2024-11-14 at 11 15 42" src="https://github.com/user-attachments/assets/a032e366-007f-46be-9762-f688c824a696">

Still have this issue with the pending button.

https://github.com/user-attachments/assets/ea3afca1-67a6-468c-91e4-086a4cbc018c

And got into this state where my vote wasn't showing and didn't proceed to the next state.

<img width="1082" alt="Screenshot 2024-11-14 at 11 11 43" src="https://github.com/user-attachments/assets/78f96b6b-6907-495e-8d8b-3b22cd14088d">

iamsamgibbs reviewed (changes requested):
Thanks for your continued work on this! It is getting closer with each change.

I still hit a couple of instances where the motion widget didn't update properly to the next stage. But I also hit the same issue when testing a funding motion too, so maybe we just try to fix these as a separate issue later across all the motions.

<img width="1277" alt="Screenshot 2024-11-18 at 11 56 38" src="https://github.com/user-attachments/assets/e06228ee-697d-4d76-9a5b-5bf51fc2466b">

I had this issue when finalizing where the "Make payout" button was already pending, but did not trigger the saga:

<img width="1279" alt="Screenshot 2024-11-18 at 11 58 30" src="https://github.com/user-attachments/assets/6ed40a19-c0f9-4732-be40-e438bc0f0cc6">

https://github.com/user-attachments/assets/c581ed1f-74c5-440e-b1e5-002a5fd06652

The stake amount is also not resetting properly after submitting a stake:

<img width="1278" alt="Screenshot 2024-11-18 at 11 56 09" src="https://github.com/user-attachments/assets/266e5d69-7147-4963-97d9-007fd0300336">

It is missing this state where the form closes back up like in the funding motion:

https://github.com/user-attachments/assets/e5da80dd-c9c3-4dba-987e-af461d4f7cad

I also noticed when finalizing the motion the pill will briefly display "Failed" before updating to "Passed" I think we should just show a loader until we have the correct state.

Nice job though, it is almost there.

CzarekDryl commented:
@iamsamgibbs I fixed the pending button; it shouldn't appear anymore. The rest of the issues you mentioned don't work in other motions either, including funding. I think something 

https://github.com/user-attachments/assets/a1a471fe-8da9-4c70-bedd-3c20a560c03d

`I also noticed when finalizing the motion the pill will briefly display "Failed" before updating to "Passed" I think we should just show a loader until we have the correct state.` there is already issue for this - https://github.com/JoinColony/colonyCDapp/issues/3390

jakubcolony reviewed (changes requested):
Overall it is pretty functional, I was able to complete the flow a few times without issues.

<img width="397" alt="image" src="https://github.com/user-attachments/assets/f8e4c109-e4d7-4daa-a92b-e2c50f773164">
<img width="388" alt="image" src="https://github.com/user-attachments/assets/4e6166af-c475-4cdd-bee0-e12b643b0be6">
<img width="390" alt="image" src="https://github.com/user-attachments/assets/9140b477-3c02-49c3-9717-f310b0751bc6">
<img width="397" alt="image" src="https://github.com/user-attachments/assets/8f784e1e-3cb3-4bb0-b510-bc56efa269a3">
<img width="391" alt="image" src="https://github.com/user-attachments/assets/7bce8f8a-328a-4579-8b2f-364b93d65965">
<img width="389" alt="image" src="https://github.com/user-attachments/assets/288214a0-a211-492e-b724-6390a4a93a68">

After finalizing the motion, I can see the Claimed label flashing:

https://github.com/user-attachments/assets/a8ad40f6-b982-402f-8908-cc0e029d1be4

I also left a commend around waiting for payouts after finalizing the motion, which is causing the console errors Sam first reported. It also causes the Make payment button to be disabled for the first 30 seconds after finalizing.

<img width="699" alt="image" src="https://github.com/user-attachments/assets/e5ed6725-427d-44dc-bb9e-4da5f6b88f75">

CzarekDryl commented:
@jakubcolony I've fixed code issues that you mentioned. After finalizing motion nothing is flashing for me, maybe rebase fixed this. Please check if it's working fine for you too.

https://github.com/user-attachments/assets/724ec751-d08d-48e9-b753-8e7c2666391c

jakubcolony reviewed (dismissed):
The Claim label flashing is no longer reproducible as the expenditure step changes to payment before it would be shown so we can count that as resolved

There are still a few noticeable flashes of the Claim now/Claim in... label next to each payment:

https://github.com/user-attachments/assets/32914f4f-eb06-4e66-bd3f-5d2f8071a44f

Apart from that it's looking good @CzarekDryl, nice work 👍

CzarekDryl commented:
@jakubcolony I had this problem only once in a few tests; I don't see where the error might be.


https://github.com/user-attachments/assets/1ba291f7-eba6-4f3f-a730-64aba10100c2

https://github.com/user-attachments/assets/c092c6e0-3721-4dfc-b61b-61e818adc6f3

jakubcolony reviewed (dismissed):
I'm also seeing the timer countdown flashing repeatedly:

https://github.com/user-attachments/assets/b4fc1ef6-7a3f-46e4-b583-22c43f773cbd

After checking out `master`, the issue persist, so it's not something introduced in this PR. I opened #3850 to address that.

This PR is good to go, nice work overall @CzarekDryl 👍

arrenv reviewed (changes requested):
@CzarekDryl Thanks for your persistence on this feature, unfortunately, I ran into some breaking issues.

I ran into issues with Opposing and Failed motions.

1. When the timer ended when fully opposed, nothing happened, you could still Support, but that failed.
![image](https://github.com/user-attachments/assets/66f362c8-32ff-4596-85fe-ffa97fb71251)

2. I had to reload the page for it to update, then, in that instance, I was not able to claim my stakes from the previous motion.
![image](https://github.com/user-attachments/assets/4ce22382-c6b4-47e4-87a5-354973b33a66)

3. I noticed was that the "Payment" was not triggered automatically on finalization, is this doable?
![image](https://github.com/user-attachments/assets/cfb84639-6711-4f4b-b08e-705b174e6a35)



Other then that, voting and staking worked.

![image](https://github.com/user-attachments/assets/842a100e-8f7e-4cdc-a44e-1d2c2d987880)

![image](https://github.com/user-attachments/assets/5f7ad376-72fb-4879-ad0c-88a226214a4e)

CzarekDryl commented:
@arrenv I corrected the first and second points, but in the second one, it works like this in every motion. If the motion is only "opposed," it is not possible to claim the funds. Third point will be fixed in other issue. 

Here is example of simple payment fail:
![image](https://github.com/user-attachments/assets/c1e1cf84-9591-4468-a7b7-a83ff83e1159)
![image](https://github.com/user-attachments/assets/eb9835b7-293d-47e0-8b9c-e4068915bba0)

jakubcolony commented:
Just a note this PR will most likely have a conflict with #3922 as the designs have been updated (as per #3776).

arrenv reviewed (approved):
@jakubcolony Thank you for the fixes, the issues seem resolved now. All looks good from my testing with this PR.

jakubcolony reviewed (approved):
Nice one @CzarekDryl, looks like there's a few more conflicts to resolve so feel free to ping me once that's done so I can quickly re-approve.

CzarekDryl commented:
@arrenv @jakubcolony PR rebased

bassgeta reviewed (approved):
Nice going, thanks for your hard work on this :muscle: 
The code is straight forward and makes sense, and best of all, it also works exactly as it should :sunglasses: 
Reputation shows up as a decision method :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/8436c16f-e6be-4697-862a-a9e3f5113547)
Gets created  :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/25c0f21d-57d3-4cd1-832c-c6ed3a9bf463)
Failing is handled correctly :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/adbc5d13-4b54-4076-b97e-980918b7b625)
The list shows up and you can switch between funding requests :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/99dc5dcb-5435-489f-a832-b889a385b3a2)
Finalization also works :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/54497c09-c5b0-453e-a88d-e2d176e39151)

Voting also works :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/2d34ede3-5ed9-40b5-ba27-eae701656972)
Failed vote rejects the motion and you can release again :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/b11a96c4-3782-4ba9-b960-516a3a8263bc)
Support works :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/29454b61-05eb-408c-9367-0e8d3ee3db03)
![image](https://github.com/user-attachments/assets/fb036582-adf4-4d1e-b074-685348d3c1eb)

jakubcolony reviewed (commented):
Good to go, nice work @CzarekDryl 👍 

The styling for a single item in the release requests list was updated in the meantime, but I opened a separate follow-up issue: #3967.
<img width="325" alt="image" src="https://github.com/user-attachments/assets/0b33c772-e1d6-486c-8b94-60c097df7426" />


<img width="514" alt="image" src="https://github.com/user-attachments/assets/fc1f3a3b-357b-462a-a9cd-0bb063c70068" />

<img width="381" alt="image" src="https://github.com/user-attachments/assets/3b6b833c-c355-4c3b-9cc9-961d67521c84" />

<img width="381" alt="image" src="https://github.com/user-attachments/assets/05d5e399-eccc-401f-b645-edd8beaed3de" />

<img width="385" alt="image" src="https://github.com/user-attachments/assets/12177756-5a70-4d13-b2af-87120ea88b00" />

<img width="379" alt="image" src="https://github.com/user-attachments/assets/06416232-a2e6-45c0-9da6-6a840170131f" />

jakubcolony reviewed (approved):
Accidentally commented instead of approving yesterday
